src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -167,15 +167,37 @@ private async Task<FormulaValue> TryHandleSet(CallNode node, EvalVisitorContext
 
             var newValue = await arg1.Accept(this, context).ConfigureAwait(false);
 
-            if (arg0.IRContext.IsMutation && arg0 is RecordFieldAccessNode rfan)
+            if (arg0.IRContext.IsMutation)
             {
-                var arg0value = await rfan.From.Accept(this, context).ConfigureAwait(false);
-                
-                if (arg0value is RecordValue rv)
+                if (arg0 is RecordFieldAccessNode rfan)
                 {
-                    rv.ShallowCopyFieldInPlace(rfan.Field);
-                    rv.UpdateField(rfan.Field, newValue);
-                    return node.IRContext.ResultType._type.Kind == DKind.Boolean ? FormulaValue.New(true) : FormulaValue.NewVoid();
+                    var arg0value = await rfan.From.Accept(this, context).ConfigureAwait(false);
+
+                    if (arg0value is RecordValue rv)
+                    {
+                        rv.ShallowCopyFieldInPlace(rfan.Field);
+                        rv.UpdateField(rfan.Field, newValue);
+                        return node.IRContext.ResultType._type.Kind == DKind.Boolean ? FormulaValue.New(true) : FormulaValue.NewVoid();
+                    }
+                    else
+                    {
+                        return CommonErrors.UnreachableCodeError(node.IRContext);
+                    }
+                }
+                else if (arg0 is BinaryOpNode bon && bon.Op == BinaryOpKind.DynamicGetField)
+                {
+                    var arg0value = await bon.Left.Accept(this, context).ConfigureAwait(false);
+
+                    if (arg0value is UntypedObjectValue uov)
+                    {
+                        TextLiteralNode textLiteralNode = (TextLiteralNode)bon.Right;
+                        UntypedObjectValue untypedObjectValue = (UntypedObjectValue)newValue;
+
+                        // !!!TODO What if property name does not exist?

```

Comment:
>// !!!TODO What if property name does not exist? [](http://example.com/codeflow?start=24&length=48)

let the UO impl worry about this. 
May be we just allow expando. 

Key thing is that it can fail.  Let's just say host can throw a [https://github.com/microsoft/Power-Fx/blob/main/src/libraries/Microsoft.PowerFx.Interpreter/Exceptions/CustomFunctionErrorException.cs](https://github.com/microsoft/Power-Fx/blob/main/src/libraries/Microsoft.PowerFx.Interpreter/Exceptions/CustomFunctionErrorException.cs) and we catch that and convert to an errorValue.